src/WebJobs.Script.WebHost/Security/BlobStorageSecretsRepository.cs

```diff
@@ -0,0 +1,163 @@
+            _fileWatcher.Changed += OnChanged;
+
+            _secretsBlobPath = siteHostName.ToLowerInvariant(); // Config.ScriptSettingsManager.Instance.AzureWebsiteDefaultSubdomain;
+            _hostSecretsBlobPath = string.Format("{0}/{1}", _secretsBlobPath, ScriptConstants.HostMetadataFileName);
+
+            _accountConnectionString = accountConnectionString;
+            CloudStorageAccount account = CloudStorageAccount.Parse(_accountConnectionString);
+            CloudBlobClient client = account.CreateCloudBlobClient();
+
+            _blobContainer = client.GetContainerReference(_secretsContainerName);
+            _blobContainer.CreateIfNotExists();
+            //await _blobContainer.CreateIfNotExistsAsync();        
+        }
+
+        public event EventHandler<SecretsChangedEventArgs> SecretsChanged;
+
+        private string GetSecretsBlobFilePath(ScriptSecretsType secretsType, string functionName = null)
+        {
+            return secretsType == ScriptSecretsType.Host
+                ? _hostSecretsBlobPath
+                : string.Format("{0}/{1}", _secretsBlobPath, string.Format(CultureInfo.InvariantCulture, "{0}.json", functionName.ToLowerInvariant()));
+        }
+
+        private string GetSecretsFilePath(ScriptSecretsType secretsType, string functionName = null)
+        {
+            return secretsType == ScriptSecretsType.Host
+                ? _hostSecretsMarkerPath
+                : GetFunctionSecretsFilePath(functionName);
+        }
+
+        private string GetFunctionSecretsFilePath(string functionName)
+        {
+            string secretFileName = string.Format(CultureInfo.InvariantCulture, "{0}.json", functionName.ToLowerInvariant());
+            return Path.Combine(_secretsFileMarkersPath, secretFileName);
+        }
+
+        private void OnChanged(object sender, FileSystemEventArgs e)
+        {
+            var changeHandler = SecretsChanged;
+            if (changeHandler != null)
+            {
+                var args = new SecretsChangedEventArgs { SecretsType = ScriptSecretsType.Host };
+
+                if (string.Compare(Path.GetFileName(e.FullPath), ScriptConstants.HostMetadataFileName, StringComparison.OrdinalIgnoreCase) != 0)
+                {
+                    args.SecretsType = ScriptSecretsType.Function;
+                    args.Name = Path.GetFileNameWithoutExtension(e.FullPath).ToLowerInvariant();
+                }
+
+                changeHandler(this, args);
+            }
+        }
+
+        public async Task<string> ReadAsync(ScriptSecretsType type, string functionName)
+        {
+            string filePath = GetSecretsFilePath(type, functionName);
+            string blobPath = GetSecretsBlobFilePath(type, functionName);
+            string secretsContent = null;
+
+            // load the secrets from blob
+            if (_blobContainer.GetBlockBlobReference(blobPath).Exists())
```

Comment:
Get blob reference and keep that to avoid the duplication below and use `ExistsAsync` instead (which gives us the nice benefit of avoiding a non-async path)